lib/token-store.js

```diff
@@ -8,196 +8,1110 @@
      */
-    api.getTokenAfter = function(node, skip) {
-        return tokens[lastTokenIndex(node) + (skip || 0) + 1];
-    };
+    constructor(cursor, predicate) {
+        super(cursor);
+        this.predicate = predicate;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+        while (super.moveNext()) {
+            if (this.predicate(this.current)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+        while (super.movePrev()) {
+            if (this.predicate(this.current)) {
+                return true;
+            }
+        }
+        return false;
+    }
+}
+
+/**
+ * The decorative cursor which ignores the first few tokens.
+ * @private
+ */
+class SkipCursor extends DecorativeCursor {
 
     /**
-     * Gets all tokens that are related to the given node.
-     * @param {ASTNode} node The AST node.
-     * @param {int} [beforeCount=0] The number of tokens before the node to retrieve.
-     * @param {int} [afterCount=0] The number of tokens after the node to retrieve.
-     * @returns {Token[]} Array of objects representing tokens.
+     * Initializes this cursor.
+     * @param {Cursor} cursor - The cursor to be decorated.
+     * @param {number} count - The count of tokens this cursor skips.
      */
-    api.getTokens = function(node, beforeCount, afterCount) {
-        return get(
-            starts[node.range[0]] - (beforeCount || 0),
-            lastTokenIndex(node) + (afterCount || 0) + 1
-        );
-    };
+    constructor(cursor, count) {
+        super(cursor);
+        this.count = count;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+        while (this.count > 0) {
+            this.count -= 1;
+            if (!super.moveNext()) {
+                return false;
+            }
+        }
+        return super.moveNext();
+    }
+
+    /** @inheritdoc */
+    movePrev() {
+        while (this.count > 0) {
+            this.count -= 1;
+            if (!super.movePrev()) {
+                return false;
+            }
+        }
+        return super.movePrev();
+    }
+}
+
+/**
+ * The decorative cursor which limits the number of tokens.
+ * @private
+ */
+class LimitCursor extends DecorativeCursor {
 
     /**
-     * Gets the first `count` tokens of the given node's token stream.
-     * @param {ASTNode} node The AST node.
-     * @param {int} [count=0] The number of tokens of the node to retrieve.
-     * @returns {Token[]} Array of objects representing tokens.
+     * Initializes this cursor.
+     * @param {Cursor} cursor - The cursor to be decorated.
+     * @param {number} count - The count of tokens this cursor iterates.
+     */
+    constructor(cursor, count) {
+        super(cursor);
+        this.count = count;
+    }
+
+    /** @inheritdoc */
+    moveNext() {
+        if (this.count > 0) {
+            this.count -= 1;
+            return super.moveNext();
```

Comment:
It looks like after calling `moveNext` enough times, `movePrev` won't work at all. 

```js
for (let i = 0; i < count; i++) {
    this.moveNext();
}
this.movePrev(); // => false
```

I would expect that after using `movePrev`, the cursor would be at index `count - 2`.